File: accbot-android/app/src/main/java/com/accbot/dca/domain/usecase/ResolvePendingTransactionsUseCase.kt
package com.accbot.dca.domain.usecase

import android.util.Log
import com.accbot.dca.data.local.CredentialsStore
import com.accbot.dca.data.local.DcaDatabase
import com.accbot.dca.data.local.UserPreferences
import com.accbot.dca.domain.model.TransactionSide
import com.accbot.dca.domain.model.TransactionStatus
import com.accbot.dca.exchange.ExchangeApiFactory
import com.accbot.dca.service.NotificationService
import javax.inject.Inject

/**
 * Resolves PENDING/PARTIAL transactions by querying exchange APIs for fill details.
 *
 * Two scenarios produce resolvable rows:
 *  - BUY: Kraken/Coinbase sometimes return PENDING at place-order time because fill
 *    details weren't available within the initial 3-second polling window.
 *  - SELL: limit sell orders are always PENDING until (partially) filled or cancelled.
 *
 * Uses the guarded [TransactionDao.updateResolvedTransaction] UPDATE so a concurrent
 * user cancel (which sets status=FAILED) is never clobbered.
 *
 * Fires a system notification for each SELL that transitions to COMPLETED in this run.
 */
class ResolvePendingTransactionsUseCase @Inject constructor(
    private val database: DcaDatabase,
    private val credentialsStore: CredentialsStore,
    private val exchangeApiFactory: ExchangeApiFactory,
    private val userPreferences: UserPreferences,
    private val notificationService: NotificationService
) {
    suspend operator fun invoke(): Int {
        val pendingTransactions = database.transactionDao().getResolvablePendingTransactions()
        if (pendingTransactions.isEmpty()) return 0

        val isSandbox = userPreferences.isSandboxMode()
        var resolvedCount = 0

        for (tx in pendingTransactions) {
            try {
                @Suppress("DEPRECATION")
                val credentials = if (tx.connectionId != null) {
                    credentialsStore.getCredentials(tx.connectionId, isSandbox)
                } else {
                    credentialsStore.getCredentials(tx.exchange, isSandbox)
                } ?: continue
                val api = exchangeApiFactory.create(credentials)
                val orderId = tx.exchangeOrderId ?: continue

                val result = api.getOrderStatus(orderId, tx.crypto, tx.fiat) ?: continue

                val newPrice = result.avgFillPrice ?: tx.price
                val rows = database.transactionDao().updateResolvedTransaction(
                    id = tx.id,
                    newStatus = result.status,
                    cryptoAmount = result.filledCryptoAmount,
                    fiatAmount = result.filledFiatAmount,
                    price = newPrice,
                    fee = result.fee ?: tx.fee
                )
                if (rows > 0) {
                    resolvedCount++
                    if (tx.side == TransactionSide.SELL && result.status == TransactionStatus.COMPLETED) {
                        try {
                            notificationService.showSellFilledNotification(
                                crypto = tx.crypto,
                                cryptoAmount = result.filledCryptoAmount,
                                fiatAmount = result.filledFiatAmount,
                                fiat = tx.fiat,
                                price = newPrice,
                                transactionId = tx.id,
                                planId = tx.planId ?: 0,
                                exchange = tx.exchange,
                                connectionId = tx.connectionId
                            )
                        } catch (e: Exception) {
                            Log.w(TAG, "Sell-filled notification failed for tx ${tx.id}", e)
                        }
                    }
                }
            } catch (e: Exception) {
                Log.w(TAG, "Failed to resolve pending transaction ${tx.id}", e)
            }
        }

        if (resolvedCount > 0) {
            Log.d(TAG, "Resolved $resolvedCount/${pendingTransactions.size} pending transactions")
        }
        return resolvedCount
    }

    companion object {
        private const val TAG = "ResolvePendingTx"
    }
}

File: accbot-ios/AccBot/Domain/UseCase/ResolvePendingTransactionsUseCase.swift
import Foundation
import os

/// Resolves PENDING transactions by querying exchange APIs for fill details.
///
/// When exchanges return PENDING status (fill details not available within the initial polling window),
/// the transaction is saved with cryptoAmount=0. This use case finds those transactions
/// and queries the exchange for actual fill details, updating them to COMPLETED.
final class ResolvePendingTransactionsUseCase {
    private let database: DcaDatabase
    private let credentialsStore: CredentialsStore
    private let exchangeApiFactory: ExchangeApiFactory
    private let userPreferences: UserPreferences
    private let logger = Logger(subsystem: "com.accbot.dca", category: "ResolvePendingTx")

    init(
        database: DcaDatabase,
        credentialsStore: CredentialsStore,
        exchangeApiFactory: ExchangeApiFactory,
        userPreferences: UserPreferences
    ) {
        self.database = database
        self.credentialsStore = credentialsStore
        self.exchangeApiFactory = exchangeApiFactory
        self.userPreferences = userPreferences
    }

    /// Resolve all pending transactions. Returns the number of resolved transactions.
    @discardableResult
    func invoke() async -> Int {
        do {
            let pendingTxs = try database.transactionDao.getPendingTransactions()
                .filter { $0.exchangeOrderId != nil }

            guard !pendingTxs.isEmpty else { return 0 }

            let isSandbox = userPreferences.isSandboxMode()
            var resolvedCount = 0

            for tx in pendingTxs {
                guard let orderId = tx.exchangeOrderId else { continue }
                guard let credentials = credentialsStore.get(for: tx.exchange, isSandbox: isSandbox) else { continue }

                do {
                    let api = exchangeApiFactory.create(credentials: credentials)
                    guard let filled = await api.getOrderStatus(orderId: orderId) else { continue }

                    let updated = Transaction(
                        id: tx.id,
                        planId: tx.planId,
                        exchange: tx.exchange,
                        crypto: tx.crypto,
                        fiat: tx.fiat,
                        fiatAmount: filled.fiatAmount,
                        cryptoAmount: filled.cryptoAmount,
                        price: filled.price,
                        fee: filled.fee,
                        feeAsset: filled.feeAsset,
                        status: filled.status,
                        exchangeOrderId: tx.exchangeOrderId,
                        executedAt: tx.executedAt
                    )
                    try database.transactionDao.update(updated)
                    resolvedCount += 1

                    logger.info("Resolved pending tx \(tx.id): \(updated.cryptoAmount) \(tx.crypto) for \(updated.fiatAmount) \(tx.fiat)")
                } catch {
                    logger.warning("Failed to resolve pending tx \(tx.id): \(error.localizedDescription)")
                }
            }

            if resolvedCount > 0 {
                logger.info("Resolved \(resolvedCount)/\(pendingTxs.count) pending transactions")
            }

            return resolvedCount
        } catch {
            logger.warning("Failed to fetch pending transactions: \(error.localizedDescription)")
            return 0
        }
    }
}
